### PR TITLE
Htmlunit tuning

### PR DESCRIPTION
**Pull request by rschwietzke**
HtmlUnit tuning regarding speed and memory. Current success, more on the memory side.
Please run the full test program, there might be something hidden. The small one looked ok.

### NEW (JDK 17)
```
Benchmark                                    Mode  Cnt        Score   Error   Units
HtmlUnitBenchmark.simple                     avgt    2  9639101.018           ns/op
HtmlUnitBenchmark.simple:gc.alloc.rate.norm  avgt    2  7612275.912            B/op
```

### OLD (JDK 17)
```
Benchmark                                    Mode  Cnt         Score   Error   Units
HtmlUnitBenchmark.simple                     avgt    2  10527569.765           ns/op
HtmlUnitBenchmark.simple:gc.alloc.rate.norm  avgt    2   9213538.844            B/op
``` 

### NEW (JDK 17)
```
Benchmark                                    Mode  Cnt       Score   Error   Units
HtmlUnitBenchmark.simple                     avgt    2  397427.602           ns/op
HtmlUnitBenchmark.simple:gc.alloc.rate.norm  avgt    2  383380.466            B/op
```
### OLD (JDK 17)
```
Benchmark                                    Mode  Cnt       Score   Error   Units
HtmlUnitBenchmark.simple                     avgt    2  411880.438           ns/op
HtmlUnitBenchmark.simple:gc.alloc.rate.norm  avgt    2  454900.410            B/op
```

### CODE REVIEW

File: src/main/java/org/htmlunit/util/StringUtils.java [hunk 1]
```diff
@@ -49,8 +49,8 @@ public final class StringUtils {
                                 + "\\s*((0|[1-9]\\d?|100)(.\\d*)?)%\\s*\\)");
     private static final Pattern ILLEGAL_FILE_NAME_CHARS = Pattern.compile("\\\\|/|\\||:|\\?|\\*|\"|<|>|\\p{Cntrl}");
 
-    private static final Map<String, String> CamelizeCache_ = new HashMap<>();
-    private static final Map<String, String> RootLowercaseCache_ = new HashMap<>();
+    private static final Map<String, String> CamelizeCache_ = new ConcurrentHashMap<>();
+    private static final Map<String, String> RootLowercaseCache_ = new ConcurrentHashMap<>();
```
Comment (rbri) [inline src/main/java/org/htmlunit/util/StringUtils.java:53]: RootLowercaseCache_  seems to be no longer used - does having this case makes no sense?

File: src/test/java/org/htmlunit/util/RandomUtils.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.htmlunit.util;
+
+import java.util.Random;
+
+public final class RandomUtils {
```
Comment (rbri) [inline src/test/java/org/htmlunit/util/RandomUtils.java:19]: please add author

**Comment by rbri** [comment]
https://jenkins.wetator.org/job/htmlunit/view/change-requests/job/PR-715/

the failing tests are my ones - i'm still in the middle of a huge refactoring

**Comment by rbri** [comment]
https://jenkins.wetator.org/job/HtmlUnit%20-%20Library%20Tests/view/change-requests/job/PR-715/

lets see if they also pass

**Comment by rbri** [comment]
@shapiroronny is this something you are also interested in? Will be great to see if all the hard work from @rschwietzke has some impact for your use case

**Comment by rbri** [comment]
https://jenkins.wetator.org/job/HtmlUnit%20-%20Library%20Tests/view/change-requests/job/PR-715/2/

Looks great

**Comment by shapiroronny** [comment]
> @shapiroronny is this something you are also interested in? Will be great to see if all the hard work from @rschwietzke has some impact for your use case

Hopefully yes. We're disabling javascript for our case and don't do any navigations and very little DOM mutation so not sure which flows are optimized. If initial parsing is faster, that's huge for us.

In any case, we haven't bumped in a while so we're happy upgrading in steps in case this was planned to go in the next major version with Computed Css fixes.

**Comment by rbri** [comment]
> Hopefully yes. We're disabling javascript for our case and don't do any navigations and very little DOM mutation so not sure which flows are optimized. If initial parsing is faster, that's huge for us.

Great; that was my understanding of your use case and of @rschwietzke 's goal.
Will merge that soon and we will see....

File: src/main/java/org/htmlunit/util/StringUtils.java (hunk 1 above)
Comment (rbri) [reply src/main/java/org/htmlunit/util/StringUtils.java:53]: found the answer in the commit comment

**Pull request merged**